app/documents.js

```diff
@@ -172,7 +173,12 @@ export class Document extends Base {
       const doc = await this.buildDocument(model, i);
       update();
       await delay(0);
-      this.documents[model.name].push(doc);
+      // only push the document if the key is not already taken
+      /* eslint no-underscore-dangle: ["error", { "allow": ["__name", "__key"] }] */
+      if (!built_docs_map[doc.__key]) {
```

Comment:
What you have here is a great solution to this problem. But it can be simplified a little bit and also reduce memory consumption.

The first thing is changing `built_docs_map[doc.__key] === undefined` to be just `built_docs_map[doc.__key]`. In JS `undefined` is falsy so we can use that to our advantage.
The second thing is to change `built_docs_map[doc.__key] = doc;` to be `built_docs_map[doc.__key] = true;`. This is so that we aren't storing the potentially large document in multiple places.

```js
if (built_docs_map[doc.__key]) {
  built_docs_map[doc.__key] = true;
  this.documents[model.name].push(doc);
}
```

Reply:
Ok very nice. I changed it to true. 

Also went with `!built_docs_map[doc.__key]` since I want to push for the case that it's not yet defined